File: matlab/infinite.m
%%%%%%%%%%%%%%%%INFINITE PATH GENERATOR%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Generation of a long path with a huge step number entered by the user

function y = infinite(ProteinName, path_length)
y = 0;

% --- select potential type (kept as in original) ---
pottype = '1';
if pottype=='1'
    Type = '_atomistic'; tip = 'atomistic';
elseif pottype=='2'
    Type = '_BJ';        tip = 'BJ';
elseif pottype=='3'
    Type = '_TD';        tip = 'TD';
elseif pottype=='4'
    Type = '_dist';      tip = 'dist';
elseif pottype=='5'
    Type = '_markov';    tip = 'markov';
else
    Type = '_atomistic'; tip = 'atomistic';
end

% --- read inputs (.cor and probability .out) ---
File = strcat(ProteinName, '.cor');
if ~exist(File, 'file')
    error('Missing file: %s', File);
end
coor = dlmread(File);  % input, output of the readpdb.f

File = strcat(ProteinName, Type, '.out');
if ~exist(File, 'file')
    error('Missing file: %s', File);
end
normalized = dlmread(File); % probability matrix

% --- starting indices and path length (kept default behavior) ---
baslangic  = 1;  % starting residue number
baslangic1 = 1;  % chain number

% robust parse for Octave and MATLAB (keeps your original intent)
if isnumeric(path_length)
    pathlength = path_length;
else
    pathlength = str2num(path_length); %#ok<ST2NM>
end
if isempty(pathlength) || ~isscalar(pathlength) || pathlength<1
    error('Invalid path_length. Provide a positive scalar or numeric string.');
end
pathlength = floor(pathlength);

fromres = [num2str(baslangic) ',' num2str(baslangic1)]; %#ok<NASGU>

m = size(normalized);
resno = m(1);
upper = 1;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% find initial row index from coor (as in original)
initial = 1;
for i = 1:resno   % correcting the input residue numbers
    if coor(i,10)==baslangic1 && coor(i,1)==baslangic  % to obtain row numbers
        initial = i;
    end
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% build newmtrx with header row/col (as in original)
newmtrx = zeros(resno+1, resno+1);
for i = 1:resno+1
    newmtrx(1,i) = i-1;
    newmtrx(i,1) = i-1;
end
for i = 1:resno
    for j = 1:resno
        newmtrx(i+1,j+1) = normalized(i,j);
    end
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% NEW MATRIX GENERATION (kept identical; minor safety for sums)
if str2num(pottype)==1 %#ok<ST2NM>
    cap = ones(resno,resno); %#ok<NASGU>
    sumv = 0;
    for i = 2:resno+1
        for j = 2:resno+1
            sumv = newmtrx(i,j) + sumv;
        end
        if sumv==0
            % avoid division by zero; keep row as zeros (no outgoing probs)
            % preserves original behavior as closely as possible
        else
            for j = 2:resno+1
                newmtrx(i,j) = newmtrx(i,j)/sumv;
                if i==j, newmtrx(i,j) = 1; end
            end
        end
        minimum = min(newmtrx(i,2:end));
        for j = 2:resno+1
            if minimum==newmtrx(i,j)
                newmtrx(i,j) = 0;
            end
            if i==j
                newmtrx(i,j) = 0;
            end
        end
        sumv = 0;
        for j = 2:resno+1
            sumv = newmtrx(i,j) + sumv;
        end
        if sumv==0
            % keep row zeros if no transitions remain
        else
            for j = 2:resno+1
                newmtrx(i,j) = newmtrx(i,j)/sumv;
                if i==j, newmtrx(i,j) = 0; end
            end
        end
        sumv = 0;
    end
    lower = 0;
else
    cap = ones(resno,resno); %#ok<NASGU>
    sumv = 0;
    for i = 2:resno+1
        for j = 2:resno+1
            sumv = newmtrx(i,j) + sumv;
        end
        average = sumv / resno;
        if sumv==0
            % leave row zeros
        else
            for j = 2:resno+1
                newmtrx(i,j) = newmtrx(i,j)/sumv;
                if i==j, newmtrx(i,j) = 1; end
            end
        end
        for j = 2:resno+1
            if newmtrx(i,j) < average
                newmtrx(i,j) = 0;
            end
            if i==j
                newmtrx(i,j) = 0;
            end
        end
        sumv = 0;
        for j = 2:resno+1
            sumv = newmtrx(i,j) + sumv;
        end
        counter = 0; %#ok<NASGU>
        if sumv==0
            % keep zeros
        else
            for j = 2:resno+1
                newmtrx(i,j) = newmtrx(i,j)/sumv;
                if i==j
                    newmtrx(i,j) = 0;
                end
                if newmtrx(i,j) ~= 0
                    counter = counter+1;
                    cap(i-1,counter) = newmtrx(i,j); %#ok<AGROW>
                end
            end
        end
        sumv = 0;
    end
    lower = 0;
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% PATH GENERATION (kept same; just minor safety on empty transitions)
runs = 0;          % counts the number of paths generated
prob = 0;          % probability cost accumulator (unused but kept)
taban = max(1, floor(pathlength/10));
permtrx = newmtrx;
runsonscreen = 0;  %#ok<NASGU>

startres = 1;         % original sets 1
runs = runs+1;
newmtrx = permtrx;
counter = 1;
sequence(counter, runs) = startres; %#ok<AGROW>
path(runs, counter) = coor(startres,1) + coor(startres,10)/10; %#ok<AGROW>

while counter ~= pathlength
    counter = counter + 1;
    j = 0;    % number of nonzero energy residues
    clear gen
    clear P
    clear Q
    P = 0;
    for i = 2:resno+1
        if newmtrx(startres+1, i) > lower && newmtrx(startres+1, i) < upper
            j = j + 1;
            P(1,j) = newmtrx(1,i);
            P(2,j) = newmtrx(startres+1,i);
        end
    end
    P = abs(P);

    sumv = 0;
    for i = 1:j
        sumv = sumv + P(2,i);
    end

    if j==0 || sumv==0
        % No outgoing transitions: stay in place (keeps original spirit without crashing)
        startres = startres;
    else
        for i = 1:j
            Q(1,i) = P(1,i);
            Q(2,i) = P(2,i)/sumv;
        end
        number = rand;
        sumv = 0;
        for i = 1:j
            gen(1,i) = Q(1,i);
            sumv = sumv + Q(2,i);
            gen(2,i) = sumv;
        end
        for i = 1:j
            if number < gen(2,i)
                startres = gen(1,i);
                break;
            end
        end
    end

    sequence(counter, runs) = startres; %#ok<AGROW>
    path(runs, counter) = coor(sequence(counter,runs),1) + coor(sequence(counter,runs),10)/10; %#ok<AGROW>

    if mod(counter, taban) == 0
        fprintf('Number of steps generated so far: %d\n', counter);
    end
end

s = size(path);
pl = s(2); % total path length
pn = s(1); %#ok<NASGU> % total number of paths

for i = 1:pl
    if max(abs(path(:, (pl+1-i)))) ~= 0
        j = pl+1-i; %#ok<NASGU>
        break
    end
end
paths = path; % original "trim" logic preserved but effectively already tight

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% WRITING PATHS
pl_str = num2str(pl);
what = strcat(ProteinName, Type, '_', pl_str, 'steps_infinite');
File = [what '.path'];
dlmwrite(File, path, '\t');

pl_num = str2num(pl_str); %#ok<ST2NM>
coor(:,1) = coor(:,1) + coor(:,10)/10;
stats = coor(:,1);
stats(:,2) = 0;
for i = 1:pl_num
    x = find(stats(:,1) == path(1,i));
    if ~isempty(x)
        stats(x,2) = stats(x,2) + 1;
    end
end

end

File: mcpath/infinite.py
# infinite.py
# Python translation of MATLAB "infinite" function
# (single long path; same logic as the original code)

import os
import numpy as np


def infinite(ProteinName: str, path_length: int, pottype: str = "1"):
    """
    Generate a long residue path using the normalized matrix in
    <ProteinName><Type>.out and coordinates in <ProteinName>.cor

    Parameters
    ----------
    ProteinName : str
        Base name of the files (no extension).
    path_length : int
        Number of steps to generate (MATLAB: pathlength).
    pottype : {'1','2','3','4','5'}
        1='atomistic', 2='BJ', 3='TD', 4='dist', 5='markov'.

    Returns
    -------
    path : np.ndarray, shape (1, L)
        Encoded path values: resID + chainID/10
    """

    # ----- map pottype -> Type suffix (as in MATLAB) -----
    type_map = {
        "1": "_atomistic",
        "2": "_BJ",
        "3": "_TD",
        "4": "_dist",
        "5": "_markov",
    }
    if pottype not in type_map:
        raise ValueError("pottype must be one of '1','2','3','4','5'")
    Type = type_map[pottype]

    # ----- load inputs -----
    cor_file = f"{ProteinName}.cor"
    out_file = f"{ProteinName}{Type}.out"

    if not os.path.isfile(cor_file):
        raise FileNotFoundError(cor_file)
    if not os.path.isfile(out_file):
        raise FileNotFoundError(out_file)

    # coor: same usage as MATLAB: col1=resID, col10=chainID
    coor = np.loadtxt(cor_file)
    normalized = np.loadtxt(out_file)

    # MATLAB defaults:
    baslangic = 1    # residue number (unused for start in MATLAB infinite)
    baslangic1 = 1   # chain number
    pathlength = int(path_length)

    # size of probability matrix
    resno = normalized.shape[0]
    assert normalized.shape == (resno, resno)

    # ----- build newmtrx (resno+1 x resno+1) with header row/col -----
    # MATLAB:
    # for i=1:resno+1
    #   newmtrx(1,i)=i-1;
    #   newmtrx(i,1)=i-1;
    # end
    newmtrx = np.zeros((resno + 1, resno + 1), dtype=float)
    for i in range(resno + 1):
        newmtrx[0, i] = i  # 0..resno
        newmtrx[i, 0] = i
    # newmtrx(i+1,j+1)=normalized(i,j)
    newmtrx[1:, 1:] = normalized

    upper = 1.0
    lower = 0.0

    # cap matrix (not used later, but we keep it for completeness)
    cap = np.ones((resno, resno), dtype=float)

    # ----- NEW MATRIX GENERATION -----
    # We follow the MATLAB loops explicitly.

    if int(pottype) == 1:
        # atomistic
        for i in range(1, resno + 1):  # MATLAB: i=2:resno+1
            # sum over columns 2..resno+1
            row_sum = 0.0
            for j in range(1, resno + 1):
                row_sum += newmtrx[i, j]

            if row_sum == 0:
                continue

            # first normalization + diagonal = 1
            for j in range(1, resno + 1):
                newmtrx[i, j] = newmtrx[i, j] / row_sum
                if i == j:
                    newmtrx[i, j] = 1.0

            # find minimum over whole row (including col 0)
            row_min = np.min(newmtrx[i, :])

            # zero minimum and diagonal
            for j in range(1, resno + 1):
                if newmtrx[i, j] == row_min:
                    newmtrx[i, j] = 0.0
                if i == j:
                    newmtrx[i, j] = 0.0

            # renormalize (excluding col 0)
            row_sum = 0.0
            for j in range(1, resno + 1):
                row_sum += newmtrx[i, j]

            if row_sum == 0:
                continue

            for j in range(1, resno + 1):
                newmtrx[i, j] = newmtrx[i, j] / row_sum
                if i == j:
                    newmtrx[i, j] = 0.0

    else:
        # BJ / TD / dist / markov
        for i in range(1, resno + 1):
            # original sum (unnormalized) over j=2..resno+1
            row_sum = 0.0
            for j in range(1, resno + 1):
                row_sum += newmtrx[i, j]

            if row_sum == 0:
                continue

            average = row_sum / resno

            # normalize, set diagonal=1
            for j in range(1, resno + 1):
                newmtrx[i, j] = newmtrx[i, j] / row_sum
                if i == j:
                    newmtrx[i, j] = 1.0

            # threshold by average (note: average from *pre*-normalization)
            for j in range(1, resno + 1):
                if newmtrx[i, j] < average:
                    newmtrx[i, j] = 0.0
                if i == j:
                    newmtrx[i, j] = 0.0

            # renormalize and fill cap
            row_sum = 0.0
            for j in range(1, resno + 1):
                row_sum += newmtrx[i, j]

            if row_sum == 0:
                continue

            counter = 0
            for j in range(1, resno + 1):
                newmtrx[i, j] = newmtrx[i, j] / row_sum
                if i == j:
                    newmtrx[i, j] = 0.0
                if newmtrx[i, j] != 0:
                    counter += 1
                    cap[i - 1, counter - 1] = newmtrx[i, j]

    # ----- PATH GENERATION -----
    # MATLAB:
    # startres=1; runs=runs+1; newmtrx=permtrx; counter=1; sequence(counter,runs)=startres; ...

    startres = 1  # 1-based residue index
    taban = max(1, pathlength // 10)
    rng = np.random.default_rng()

    sequence = np.zeros(pathlength, dtype=int)   # 1-based residue indices
    path = np.zeros((1, pathlength), dtype=float)

    # first step
    counter = 1
    sequence[counter - 1] = startres
    row0 = startres - 1  # convert to 0-based index into coor
    path[0, counter - 1] = float(coor[row0, 0]) + float(coor[row0, 9]) / 10.0

    while counter != pathlength:
        counter += 1

        # --- build P (indices & probabilities) ---
        # MATLAB:
        # j=0; P=0; for i=2:resno+1, if newmtrx(startres+1,i)>lower && <upper ...
        j = 0
        P_idx = []
        P_prob = []

        row_idx = startres  # since MATLAB row = startres+1 (1-based) => index 'startres'
        for col in range(1, resno + 1):  # 1..resno corresponds to col 2..resno+1
            val = newmtrx[row_idx, col]
            if (val > lower) and (val < upper):
                j += 1
                P_idx.append(int(newmtrx[0, col]))   # header = residue index
                P_prob.append(float(val))

        if j == 0:
            # MATLAB behavior: no update to startres; path stays at same residue
            sequence[counter - 1] = startres
        else:
            P_prob = np.array(P_prob, dtype=float)
            # Q(2,i) = P(2,i) / sum; cumulative in gen(2,i)
            s = P_prob.sum()
            if s <= 0:
                # should not happen; fall back to uniform among P_idx
                q = np.ones(j, dtype=float) / j
            else:
                q = P_prob / s

            cdf = np.cumsum(q)
            r = rng.random()
            k = int(np.searchsorted(cdf, r, side="right"))
            if k >= j:
                k = j - 1
            startres = int(P_idx[k])     # 1-based residue index

            sequence[counter - 1] = startres

        # append encoded residue to path
        row0 = startres - 1
        path[0, counter - 1] = float(coor[row0, 0]) + float(coor[row0, 9]) / 10.0

        if counter % taban == 0:
            print(f"Number of steps generated so far: {counter}")

    # ----- trim trailing zero columns (mirrors MATLAB, mostly no-op here) -----
    # MATLAB does a check for max(abs(path(:,pl+1-i)))~=0; here path has exactly
    # 'pathlength' entries, all filled, so we just keep them.

    pl = path.shape[1]

    # ----- write .path file -----
    out_name = f"{ProteinName}{Type}_{pl}steps_infinite.path"
    np.savetxt(out_name, path, fmt="%.10g", delimiter="\t")
    print(f"Saved path to: {out_name}")

    # ----- stats block (same idea as MATLAB; optional) -----
    coor_enc = coor[:, 0] + coor[:, 9] / 10.0
    stats = np.column_stack([coor_enc, np.zeros_like(coor_enc)])
    for val in path[0, :]:
        hits = np.where(np.isclose(stats[:, 0], val, atol=1e-12))[0]
        if hits.size > 0:
            stats[hits[0], 1] += 1

    return path
